core/arch/arm/plat-zynqmp/drivers/puf.c

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 Foundries.io Ltd.
+ *
+ * Jorge Ramirez-Ortiz <jorge@foundries.io>
+ */
+
+#include <initcall.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <mm/core_memprot.h>
+
+#include "puf.h"
+
+/* CSU */
+#define ISR				0x20
+#define PUF_ISR_ACC_ERROR_MASK          BIT(12)
+
+/* PUF */
+#define CSU_PUF_BASE			(CSU_BASE + 0x4000)
+#define CSU_PUF_SIZE			0x1000
+
+#define CMD				0x00
+#define CFG0				0x04
+#define CFG1				0x08
+#define SHUT				0x0C
+#define STATUS				0x10
+#define WORD				0x18
+
+#define PUF_REGISTRATION		1
+#define PUF_REGENERATION		4
+#define PUF_STOP			3
+#define PUF_RESET			6
+
+#define PUF_CFG0_DEFAULT		0x02
+#define PUF_CFG1_DEFAULT		0x0c230090
+#define PUF_SHUT_DEFAULT		0x01000100
+#define PUF_REGEN_TIME_MS		3000
+#define PUF_STATUS_SYN_WRD_RDY_TIMEOUT	500000
+#define PUF_STATUS_SYN_WRD_RDY_MASK	BIT(0)
+#define PUF_STATUS_KEY_RDY_MASK		BIT(3)
+
+#define	PUF_SYN_WORDS			141
+
+TEE_Result puf_regenerate(void)
+{
+	vaddr_t puf = core_mmu_get_va(CSU_PUF_BASE, MEM_AREA_IO_SEC,
+				      CSU_PUF_SIZE);
+	vaddr_t csu = core_mmu_get_va(CSU_BASE, MEM_AREA_IO_SEC,
+				      0x1000);
+	uint32_t status = 0;
+
+	io_write32(puf + CFG0, PUF_CFG0_DEFAULT);
+	io_write32(puf + SHUT, PUF_SHUT_DEFAULT);
+	io_write32(puf + CMD, PUF_REGENERATION);
+	mdelay(PUF_REGEN_TIME_MS);
+
+	status = io_read32(csu + ISR) & PUF_ISR_ACC_ERROR_MASK;
+	if (status) {
+		EMSG("regeneration failed");
+		return TEE_ERROR_GENERIC;
+	}
+
+	return TEE_SUCCESS;
+}
+
+void puf_reset(void)
+{
+	vaddr_t puf = core_mmu_get_va(CSU_PUF_BASE, MEM_AREA_IO_SEC,
+				      CSU_PUF_SIZE);
+
+	io_write32(puf + CMD, PUF_RESET);
+}
+
+/* TODO:
+ *  the syndrome words need to be post-processed and fused to generate a
+ *  persistent KEK
+ */
+uint32_t syn[PUF_SYN_WORDS];
```

Comment:
global variable?

Add at least static in here.

Reply:
no no, you are right, it needs to be static

Reply:
I still need to add the post-processing before merging this PR tough. will fix anyway